plugins/offers.c

```diff
@@ -111,6 +119,552 @@ static const struct plugin_hook hooks[] = {
+						 "invoice",
+						 "refund_signature",
+						 invoice->payer_key,
+						 invoice->refund_signature))
+			json_add_string(js, "warning_invoice_refund_signature_invalid",
+					"refund_signature does not match");
+	} else if (invoice->refund_for)
+		json_add_string(js, "warning_invoice_refund_missing_signature",
+				"refund_for requires refund_signature");
+
+	/* invoice_decode checked these */
+	json_add_pubkey32(js, "node_id", invoice->node_id);
+	json_add_bip340sig(js, "signature", invoice->signature);
+}
+
+static void json_add_invoice_request(struct json_stream *js,
+				     const struct tlv_invoice_request *invreq)
+{
+	if (invreq->chains)
+		json_add_chains(js, invreq->chains);
+	/* BOLT-offers #12:
+	 * - MUST fail the request if `payer_key` is not present.
+	 * - MUST fail the request if `chains` does not include (or imply) a supported chain.
+	 * - MUST fail the request if `features` contains unknown even bits.
+	 * - MUST fail the request if `offer_id` is not present.
+	 */
+	if (invreq->offer_id)
+		json_add_sha256(js, "offer_id", invreq->offer_id);
+	else
+		json_add_string(js, "warning_invoice_request_missing_offer_id",
+				"invoice_request requires offer_id");
+	if (invreq->amount)
+		json_add_amount_msat_only(js, "amount_msat",
+					  amount_msat(*invreq->amount));
+	if (invreq->features)
+		json_add_hex_talarr(js, "features", invreq->features);
+	if (invreq->quantity)
+		json_add_u64(js, "quantity", *invreq->quantity);
+
+	if (invreq->recurrence_counter)
+		json_add_u32(js, "recurrence_counter",
+			     *invreq->recurrence_counter);
+	if (invreq->recurrence_start)
+		json_add_u32(js, "recurrence_start",
+			     *invreq->recurrence_start);
+	if (invreq->payer_key)
+		json_add_pubkey32(js, "payer_key", invreq->payer_key);
+	else
+		json_add_string(js, "warning_invoice_request_missing_payer_key",
+				"invoice_request requires payer_key");
+	if (invreq->payer_info)
+		json_add_hex_talarr(js, "payer_info", invreq->payer_info);
+
+	/* BOLT-offers #12:
+	 * - if the offer had a `recurrence`:
+	 *   - MUST fail the request if there is no `recurrence_counter` field.
+	 *   - MUST fail the request if there is no `recurrence_signature` field.
+	 *   - MUST fail the request if `recurrence_signature` is not correct.
+	 */
+	if (invreq->recurrence_signature) {
+		json_add_bip340sig(js, "recurrence_signature",
+				   invreq->recurrence_signature);
+		if (invreq->payer_key
+		    && !bolt12_check_signature(invreq->fields,
+					       "invoice_request",
+					       "recurrence_signature",
+					       invreq->payer_key,
+					       invreq->recurrence_signature))
+			json_add_string(js, "warning_invoice_request_invalid_recurrence_signature",
+					"Bad recurrence_signature");
+	} else if (invreq->recurrence_counter) {
+		json_add_string(js, "warning_invoice_request_missing_recurrence_signature",
+				"invoice_request requires recurrence_signature");
+	}
+}
+
+static struct command_result *json_decode(struct command *cmd,
```

Comment:
This needs to move into the core, independently from offers.

Reply:
Well, it could also be its own plugin.  It doesn't need anything in core, but I'm in two minds here.

Reply:
Ok, we can reconcile this later again.